### PR TITLE
Allow circuit breaker settings to be runtime controllable

### PR DESCRIPTION
**Pull request by junr03**
@lyft/network-team

### CODE REVIEW

**Comment by mattklein123** [comment]
per offline conversation:
- runtime key needs to include cluster name
- need to check for max inside max call (not during init)
- use mock runtime for tests

File: source/common/upstream/resource_manager_impl.h [hunk 1]
```diff
@@ -17,10 +18,12 @@ namespace Upstream {
  */
 class ResourceManagerImpl : public ResourceManager {
 public:
-  ResourceManagerImpl(uint64_t max_connections, uint64_t max_pending_requests,
-                      uint64_t max_requests, uint64_t max_retries)
-      : connections_(max_connections), pending_requests_(max_pending_requests),
-        requests_(max_requests), retries_(max_retries) {}
+  ResourceManagerImpl(Runtime::Loader& runtime, std::string runtime_key, uint64_t max_connections,
```
Comment (mattklein123) [inline source/common/upstream/resource_manager_impl.h]: const std::string&

File: source/common/upstream/resource_manager_impl.h [hunk 2]
```diff
@@ -30,20 +33,23 @@ class ResourceManagerImpl : public ResourceManager {
 
 private:
   struct ResourceImpl : public Resource {
-    ResourceImpl(uint64_t max) : max_(max) {}
+    ResourceImpl(uint64_t max, Runtime::Loader& runtime, std::string runtime_key)
```
Comment (mattklein123) [inline source/common/upstream/resource_manager_impl.h]: const std::string&

File: source/common/upstream/upstream_impl.cc [hunk 3]
```diff
@@ -133,31 +134,37 @@ void ClusterImplBase::setHealthChecker(HealthCheckerPtr&& health_checker) {
...
+                                                               Runtime::Loader& runtime,
+                                                               std::string cluster_name,
                                                                const std::string& priority) {
   uint64_t max_connections = 1024;
   uint64_t max_pending_requests = 1024;
   uint64_t max_requests = 1024;
   uint64_t max_retries = 3;
+  std::string runtime_prefix = "circuit_breakers." + cluster_name + "." + priority + ".";
 
+  // check against config
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.cc]: delete

File: source/common/upstream/upstream_impl.h [hunk 4]
```diff
@@ -180,8 +181,10 @@ class ClusterImplBase : public Cluster,
 
 private:
   struct ResourceManagers {
-    ResourceManagers(const Json::Object& config);
-    ResourceManagerImplPtr load(const Json::Object& config, const std::string& priority);
+    ResourceManagers(const Json::Object& config, Runtime::Loader& runtime,
+                     std::string cluster_name);
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.h]: const std::string&

File: source/common/upstream/upstream_impl.h [hunk 5]
```diff
@@ -180,8 +181,10 @@ class ClusterImplBase : public Cluster,
...
+    ResourceManagerImplPtr load(const Json::Object& config, Runtime::Loader& runtime,
+                                std::string cluster_name, const std::string& priority);
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.h]: const std::string&

File: test/common/upstream/upstream_impl_test.cc [hunk 6]
```diff
@@ -137,6 +139,44 @@ TEST(StrictDnsClusterImplTest, Basic) {
   }
 }
 
+TEST(StrictDnsClusterImplTest, RuntimeResourceManager) {
```
Comment (mattklein123) [inline test/common/upstream/upstream_impl_test.cc]: delete this test, test the runtime stuff against resource manager directly.

File: source/common/upstream/resource_manager_impl.h [hunk 7]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "envoy/upstream/resource_manager.h"
+#include "envoy/runtime/runtime.h"
```
Comment (RomanDzhabarov) [inline source/common/upstream/resource_manager_impl.h]: nit: swap these two (lexicographical order)

File: source/common/upstream/upstream_impl.h [hunk 8]
```diff
@@ -8,6 +8,7 @@
 #include "envoy/upstream/health_checker.h"
 #include "envoy/upstream/load_balancer.h"
 #include "envoy/upstream/upstream.h"
+#include "envoy/runtime/runtime.h"
```
Comment (RomanDzhabarov) [inline source/common/upstream/upstream_impl.h]: very minor, same here

File: test/common/filter/tcp_proxy_test.cc [hunk 9]
```diff
@@ -7,6 +7,7 @@
 #include "test/mocks/network/mocks.h"
 #include "test/mocks/upstream/host.h"
 #include "test/mocks/upstream/mocks.h"
+#include "test/mocks/runtime/mocks.h"
```
Comment (RomanDzhabarov) [inline test/common/filter/tcp_proxy_test.cc]: order

File: source/common/upstream/resource_manager_impl.h (hunk 1 above)
Comment (junr03) [inline source/common/upstream/resource_manager_impl.h]: changed

File: source/common/upstream/resource_manager_impl.h (hunk 2 above)
Comment (junr03) [inline source/common/upstream/resource_manager_impl.h]: changed

File: source/common/upstream/upstream_impl.cc (hunk 3 above)
Comment (junr03) [inline source/common/upstream/upstream_impl.cc]: deleted

File: source/common/upstream/upstream_impl.h (hunk 4 above)
Comment (junr03) [inline source/common/upstream/upstream_impl.h]: changed

File: source/common/upstream/upstream_impl.h (hunk 5 above)
Comment (junr03) [inline source/common/upstream/upstream_impl.h]: changed

File: test/common/upstream/upstream_impl_test.cc (hunk 6 above)
Comment (junr03) [inline test/common/upstream/upstream_impl_test.cc]: delete test, and tested resource manager directly.

File: source/common/upstream/resource_manager_impl.h (hunk 7 above)
Comment (junr03) [inline source/common/upstream/resource_manager_impl.h]: moved

File: source/common/upstream/upstream_impl.h (hunk 8 above)
Comment (junr03) [inline source/common/upstream/upstream_impl.h]: moved

File: test/common/filter/tcp_proxy_test.cc (hunk 9 above)
Comment (junr03) [inline test/common/filter/tcp_proxy_test.cc]: moved

File: source/common/upstream/upstream_impl.cc [hunk 10]
```diff
@@ -133,31 +134,36 @@ void ClusterImplBase::setHealthChecker(HealthCheckerPtr&& health_checker) {
...
 
 ResourceManagerImplPtr ClusterImplBase::ResourceManagers::load(const Json::Object& config,
+                                                               Runtime::Loader& runtime,
+                                                               const std::string& cluster_name,
                                                                const std::string& priority) {
   uint64_t max_connections = 1024;
   uint64_t max_pending_requests = 1024;
   uint64_t max_requests = 1024;
   uint64_t max_retries = 3;
+  std::string runtime_prefix = "circuit_breakers." + cluster_name + "." + priority + ".";
```
Comment (RomanDzhabarov) [inline source/common/upstream/upstream_impl.cc]: nit: fmt::format("circuit_breakers.{}.{}.", cluster_name, priority); (i think reads better)

File: test/common/http/http2/conn_pool_test.cc [hunk 11]
```diff
@@ -7,6 +7,7 @@
 #include "test/mocks/http/mocks.h"
 #include "test/mocks/network/mocks.h"
 #include "test/mocks/upstream/mocks.h"
+#include "test/mocks/runtime/mocks.h"
```
Comment (RomanDzhabarov) [inline test/common/http/http2/conn_pool_test.cc]: order

File: test/common/upstream/logical_dns_cluster_test.cc [hunk 12]
```diff
@@ -4,6 +4,7 @@
 #include "test/mocks/network/mocks.h"
 #include "test/mocks/ssl/mocks.h"
 #include "test/mocks/thread_local/mocks.h"
+#include "test/mocks/runtime/mocks.h"
```
Comment (RomanDzhabarov) [inline test/common/upstream/logical_dns_cluster_test.cc]: order

File: test/common/upstream/upstream_impl_test.cc [hunk 13]
```diff
@@ -8,6 +8,7 @@
 #include "test/mocks/network/mocks.h"
 #include "test/mocks/ssl/mocks.h"
 #include "test/mocks/upstream/mocks.h"
+#include "test/mocks/runtime/mocks.h"
```
Comment (RomanDzhabarov) [inline test/common/upstream/upstream_impl_test.cc]: order

File: test/common/upstream/upstream_impl_test.cc [hunk 14]
```diff
@@ -251,7 +256,33 @@ TEST(StaticClusterImplTest, UnsupportedFeature) {
   )EOF";
 
   Json::StringLoader config(json);
-  EXPECT_THROW(StaticClusterImpl(config, stats, ssl_context_manager), EnvoyException);
+  EXPECT_THROW(StaticClusterImpl(config, runtime, stats, ssl_context_manager), EnvoyException);
+}
+
+TEST(ResourceManagerImplTest, RuntimeResourceManager) {
```
Comment (mattklein123) [inline test/common/upstream/upstream_impl_test.cc]: move into resource_manager_impl_test.cc

File: test/common/upstream/upstream_impl_test.cc [hunk 15]
```diff
@@ -251,7 +256,33 @@ TEST(StaticClusterImplTest, UnsupportedFeature) {
...
+  NiceMock<Runtime::MockLoader> runtime;
+  ResourceManagerImpl resource_manager(
+      runtime, "circuit_breakers.runtime_resource_manager_test.default.", 0, 0, 0, 0);
+
+  EXPECT_CALL(
```
Comment (mattklein123) [inline test/common/upstream/upstream_impl_test.cc]: this is not a useful test. Same for all of the ones below. You need to actually return a non-default value in your expectation, and then verify that is what is returned.

File: test/common/upstream/upstream_impl_test.cc (hunk 14 above)
Comment (junr03) [inline test/common/upstream/upstream_impl_test.cc]: moved

File: test/common/upstream/upstream_impl_test.cc (hunk 15 above)
Comment (junr03) [inline test/common/upstream/upstream_impl_test.cc]: returning non-default values in the expectation.

File: test/common/http/http2/conn_pool_test.cc (hunk 11 above)
Comment (junr03) [inline test/common/http/http2/conn_pool_test.cc]: fixed

File: test/common/upstream/logical_dns_cluster_test.cc (hunk 12 above)
Comment (junr03) [inline test/common/upstream/logical_dns_cluster_test.cc]: fixed

File: test/common/upstream/upstream_impl_test.cc (hunk 13 above)
Comment (junr03) [inline test/common/upstream/upstream_impl_test.cc]: fixed.

File: source/common/upstream/upstream_impl.cc (hunk 10 above)
Comment (junr03) [inline source/common/upstream/upstream_impl.cc]: fixed.

File: test/common/upstream/resource_manager_impl_test.cc [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+              getInteger("circuit_breakers.runtime_resource_manager_test.default.max_requests", 0U))
+      .WillOnce(Return(3U));
+  EXPECT_EQ(3U, resource_manager.requests().max());
+
+  EXPECT_CALL(runtime.snapshot_,
+              getInteger("circuit_breakers.runtime_resource_manager_test.default.max_retries", 0U))
+      .WillOnce(Return(4U));
+  EXPECT_EQ(4U, resource_manager.retries().max());
+}
+}
```
Comment (mattklein123) [inline test/common/upstream/resource_manager_impl_test.cc]: nit: newline before the line, add // Upstream

Comment (junr03) [inline test/common/upstream/resource_manager_impl_test.cc]: fixed nit

**Pull request merged**